csharp/extractor/Semmle.Util/ToolStatusPage.cs

```diff
@@ -0,0 +1,189 @@
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Serialization;
+
+namespace Semmle.Util
+{
+    /// <summary>
+    /// Represents diagnostic messages for the tool status page.
+    /// </summary>
+    public class DiagnosticMessage
```

Comment:
There are a lot of public setters.
Is it possible to re-structure the code such that we limit the accessibility of the setters?

Reply:
I refactored all of this to avoid all of the public setters in https://github.com/github/codeql/pull/12217/commits/fea29d5172a5281cf9addeb4700348af63279b95. I needed to move [the `Language` class to `Semmle.Util`](https://github.com/github/codeql/pull/12217/commits/e3762c7f93768a6f521b6054f03c7c359221533e) for that to keep the `TspSource` creation simple, but I think that makes sense since `Language` has virtually no dependencies anyway.